File: csharp_v2_pre_alpha/src/MicroServiceWorkshop/RapidsRivers/RapidsConnection.cs
/*
 * Copyright (c) 2016 by Fred George
 * May be used freely except for training; license required for training.
 */

using System.Collections.Generic;

namespace MicroServiceWorkshop.RapidsRivers
{
    public abstract class RapidsConnection
    {
        protected readonly List<IMessageListener> Listeners = new List<IMessageListener>();

        public virtual void Register(IMessageListener listener)
        {
            Listeners.Add(listener);
        }

        public abstract void Publish(string message);

        public interface IMessageListener
        {
            void HandleMessage(RapidsConnection sendPort, string message);
        }
    }
}

File: ruby_v2/lib/rapids_rivers/rapids_connection.rb
# Copyright (c) 2017 by Fred George.
# May be used freely except for training; license required for training.

# Understands the protocol to an event bus
module RapidsConnection

  def register(listener)
    listeners << listener
  end

  def received_message(message)
    listeners.each { |listener| listener.message(self, message) }
  end

  def publish(packet)
    throw "No implementation to send packets: \n\t #{packet.to_s}"
  end

  private

    def listeners
      @listeners ||= []
    end

end
